### PR TITLE
Aws parallel cluster

### PR DESCRIPTION
**Pull request by RamiyapriyaS-Deloitte**
### Description
Added the AWS ParallelCluster tutorial. 

### Assignee
*Assignees: @zbyosufzai *

## PR checklist
*Please ensure the following:*
- [x] This comment contains a description of changes (with reason).
- [x] All changes were tested.
- [x] If you've fixed a bug mention the issue number/name.
- [x] Apply approriate tags (e.g. documentation, bug)

### CODE REVIEW

File: notebooks/AWS-ParallelCluster.ipynb [hunk 1]
```diff
@@ -0,0 +1,711 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "1b5b1446",
+   "metadata": {},
+   "source": [
+    "# Snakemake on AWS ParallelCluster \n",
+    "\n",
+    "**Difficulty Level: Beginner**\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: Can you change this to "Skill Level"?

File: notebooks/AWS-ParallelCluster.ipynb [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+    "\n",
+    "## Introduction\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: Can you follow the sections listed in the tutorial Checklist? For instance this should say "Overview". link to checklist https://nih.sharepoint.com/:w:/r/sites/GRP-CIT-CloudServices/Shared%20Documents/Cloud%20Lab/Tutorial%20Development/Cloud%20Lab%20Tutorial%20Checklist.docx?d=w8435b775749543d3a4e3104aa01e47cd&csf=1&web=1&e=SugGja

Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: As per the checklist please add in Learning Objectives, Prerequisite, and Pricing sections. Some prerequisites are certain permissions that are required to use the Parallel Cluster. You can gather this from the Parallel CLuster documentation. Other prereqs or rather a good note to add would be to let the user know they will be running the commands in this tutorial via the cloud shell provided by Parallel Cluster UI.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
...Z7i/rikVGQWKA8flUYNS/NzEBefiKSUTFQ2tCi+ciQ3IxVJ6dmokYZLVaT5hVkZiE9KQ3ZOPqrKS1HdeNd+ENFbnJulvEhSeg4q6xpQU1mKojJlyDbV1yA5IRGJqZnIyilBaWkVmlrkKBHbkJCYgoSUDBSUKMckW+QNyM9KF68nB7XSUcMigAuyMxArtjUxVXlBpOaGOpQVF6Hm7ssJq9RUliNB7Ku8ImUA1ldXICkhGXGJaSgq//XYZ7WsTNyvPEw6K78EDar11tdUISVJ7Dvx3OlZBahpUL7ualkpEuITFIdNKx5/xwWG5CgtK0NJhTKCK0qLEB+XIuZVoLRE/LdYRnrNhWWVvz73s7ke+dmZYjtTkV9YJPZJNSpVL7K6QmxjYqLiOTNzVdvYKt4z6fGpOahrbIKstFCsOxFp2YXIyxPvk3gfpO0pKC5FZW0DGsR2l1eIfd8knidXLCf2SUJyOipqGKdERET012GgEqnpoTgHlYiIiIjofwADlUhNDFQiIiIiIs1goBKpiYFKRERERKQZDFQiNTFQiYiIiIg0g4FKpCYGKhERERGRZjBQidTEQCUiIiIi0gwGKpGaGKhERERERJpxK1BlDFSi+8NAJSIiIiLSDI6gEqmJgUpEREREpBkMVCI1MVCJiIiIiDSDgUqkJgYqEREREZFmMFCJ1MRAJSIiIiLSDAYqkZoYqEREREREmtEWqFUMVKL7wUAlIiIiItIMjqASqYmBSkRERESkGQxUIjUxUImIiIiINONWoMoYqET3h4FKRERERKQZHEElUhMDlYiIiIhIMxioRGpioBIRERERaQYDlUhNDFQiIiIiIs1goBKpiYFKRERERKQZDFQiNTFQiYiIiIg0g4FKpCYGKhERERGRZjBQidTEQCUiIiIi0gwGKpGaGKhERERERJpxK1BlDFSi+8NAJSIiIiLSDI6gEqmJgUpEREREpBkMVCI1MVCJiIiIiDSDgUqkJgYqEREREZFmMFCJ1MRAJSIiIiLSDAYqkZoYqEREREREmsFAJVITA5WIiIiISDMYqERqYqASEREREWkGA5VITQxUIiIiIiLNYKASqYmBSkRERESkGQxUIjUxUImIiIiINOOOQBUfsInoTxI/QK3SD1BLi/hxIiIiIiKi+9YsPlMXl5aLz9fVrf8Plac/BtP0OJwAAAAASUVORK5CYII="
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: Please add the image to the image folder in the repo instead of adding it to the notebook directly. After you add the image to the folder then in  jupyter notebook add a markdown cell that has a code like this `! [image_name](path to image in image folder)`.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
...Qiu2OSsiWMHXRJ6hvaqNzGnuFJX8GWvUTyeDMHlOHMyuwf3OgFbpiKyrXdY1johu3rTLv2euaWs/osQvcsJVryXbNHSN7JPRvcfkpjETBWDsUmZMrTvkIptFmfCY8v9R8YP62rAeDsRgp4YZMVdFmFiQSU8npxD3HY7sTIyfkTnw5bWNGn8OvpHJCO/zAnOUdl70LdFzWOL1+q9s5zwZghwQVmdetoOOcE1tO+gfldW06y/J2+qGto17nc88ITOZyUtabklKlAvZ0/qxRziiyhl1eb7n1yr5cj2Mp6Rw6qQw6N7ZXNovDy6ZJN0dJ/Zy0o7rqmpdwJymzz21DKJjk1w1zuix1lFF4HICr11da5OuftUVdfr3FVAmObkF+nqlBggCNx1G7ZLe3unxoU5q5HR8fLkopXqje3q6tF4Et/M7Dx5YtEKWeYEbLETraxqOZX0jFxpbGrx/2XMFUygGsbs2eeE6bDrM2ca4jvf8ASoF7zn09TvEKiI1gMHDphANQzj/JlRoD7Lc1AJ+104dOykZDTulQ89Wi8vuqVa/skJ0390QvX5N1bJP99ULS92xz6zsEFKOsblsBO0M10HsbXHqcKpHkE9fuCo/o+AxPPKZ8K4+8yxvf7v8HIOO6GJoPR+z76qHMMbiuBlixl+i/Djun0j4+6cYxNCb4/7wLVO/f7YxOfAY9yD7W+4Lqv/ajz8506Ln4sD/3OPMXdj7slQYdLKcS99LITkeWkJnpeZ67JlDr/z3XNMxXNgnC7HcDEFKvD2PTop24nUdbJme5Sk5hTrcPLY5Gx5avUOWbB0s9Q0tE4YBWeDN/rj4/tde54+zHb/gQOuvXtDgE9OGFRePxBoeHA+RocHwnUf150yrJe+g+0WuPbp4Jxzjb/x7GEC1TBmz+XqQT1T4HmDQD148KAJVMMwzh/PML3YApXgidTmgYPy14gOefNdFU6YVsq/3FQh77qvUh6I75bO0cN6zky/JyDWZvIGep5LLwQem3oOnxGpBO8Ygb8RpAQ+e7/jfvwdeF8+e9fy/vY+e4HvvesRvN9453r31c/+vwN/693TO88L3jlT48DnwDR49/S+v1zDxRaogBFQU98qG0Ji5UEnGO5esEoeX75VohIyda9Uw7gQmEA1jNljAnVy4EXmnj17dIjvoUOHnv7/C7dpRjt/JqcAAAAASUVORK5CYII="
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: Same comment as above for adding images

File: notebooks/AWS-ParallelCluster.ipynb [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+    "\n",
+    "### Dependencies \n",
+    "\n",
+    "* AWS ParallelCluster UI \n",
+    "\n",
+    "Please follow the installation instructions for the ParallelCluster UI provided here: [here](https://github.com/STRIDES/NIHCloudLabAWS/blob/main/docs/Install_AWSParallelCluster.md). These instructions will guide you through the necessary steps to create a CloudFormation Stack through which you can access the AWS ParallelCluster UI. \n",
+    "\n",
+    "Additionally, we urge you to check out the documents within the `docs/` folder of the repository for more bioinformatics and Gen AI tutorials.\n",
+    "\n",
+    "Once you have created the Cloud Formation Stack for the PCUI, navigate to the user interface URL. It will look like this:"
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: This is where the 'Get started' section should be.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+   "id": "0970e82b",
+   "metadata": {},
+   "source": [
+    "## Setup \n",
+    "\n",
+    "### Dependencies \n",
+    "\n",
+    "* AWS ParallelCluster UI \n",
+    "\n",
+    "Please follow the installation instructions for the ParallelCluster UI provided here: [here](https://github.com/STRIDES/NIHCloudLabAWS/blob/main/docs/Install_AWSParallelCluster.md). These instructions will guide you through the necessary steps to create a CloudFormation Stack through which you can access the AWS ParallelCluster UI. \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: You name this section 'Prerequisites'

File: notebooks/AWS-ParallelCluster.ipynb [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+   "source": [
+    "1. Switch to user and create a working directory"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b4ef6fb3",
+   "metadata": {},
+   "source": [
+    "```bash \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: I know the user is running this from the shell and not a Jupyter notebook but can you make this cell into a Code cell just to help differentiate markdown from code?

File: notebooks/AWS-ParallelCluster.ipynb [hunk 8]
```diff
@@ -0,0 +1,711 @@
...
+   "source": [
+    "3. Install conda. We will be executing Snakemake using conda. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ccabf387",
+   "metadata": {},
+   "source": [
+    "```bash\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: same comment as above

File: notebooks/AWS-ParallelCluster.ipynb [hunk 9]
```diff
@@ -0,0 +1,711 @@
...
+    "conda --version\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6c90d562",
+   "metadata": {},
+   "source": [
+    "3. Install Snakemake and the Snakemake ParallelCluster plugin. \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: same comment as above

File: notebooks/AWS-ParallelCluster.ipynb [hunk 10]
```diff
@@ -0,0 +1,711 @@
...
+    "1. Create a slurm script. This example runs prints \"Hello World\" in an output file, then appends the file with the name of the compute node the task ran on. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "33ad0a63",
+   "metadata": {},
+   "source": [
+    "```bash \n",
+    "#hello-world.slurm\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: Why are these commented out? if this is how you configure Snakemake script please explain that.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 11]
```diff
@@ -0,0 +1,711 @@
...
+   "source": [
+    "## Submitting a \"Hello World\" job to the Slurm cluster using `sbatch`\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dc8187f0",
+   "metadata": {},
+   "source": [
+    "You can submit jobs to the slurm cluster using a slurm script and the sbatch command for submission. \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: This section is a little hard to follow I think it would be great if you add something like "Now that we have set up our head node lets script our first snakemake script!"

File: notebooks/AWS-ParallelCluster.ipynb [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+    "2. Submit the job using an `sbatch` command "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4086a0f7",
+   "metadata": {},
+   "source": [
+    "```bash\n",
+    "sbatch hello-world.slurm\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: please make this into a code cell.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 13]
```diff
@@ -0,0 +1,711 @@
...
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5e062fca",
+   "metadata": {},
+   "source": [
+    "## Snakemake workflow files\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: the flow is a little weird could you move this section before the first hello world script. you can also let the user know that you are going to compare running a snakemake script using sbatch vs pcluster-slurm. please explain the difference between the two.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 14]
```diff
@@ -0,0 +1,711 @@
...
+    "\n",
+    "1. Create a Snakefile within a project directory\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ff4ed425",
+   "metadata": {},
+   "source": [
+    "```bash \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: you can make this a code cell to help differentiate from the Snakefile. Also not everyone is familiar with vi commands so either you can use the echo tip I mentioned in my email or instruct them how to edit with vi and exit out of vim.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 15]
```diff
@@ -0,0 +1,711 @@
...
+    "2. Execute the workflow using the Snakemake command, specifying `pcluster-slurm` as the executor."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c5f34921",
+   "metadata": {},
+   "source": [
+    "```bash\n",
+    "Snakemake --executor pcluster-slurm \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: please make into code cell

File: notebooks/AWS-ParallelCluster.ipynb [hunk 16]
```diff
@@ -0,0 +1,711 @@
...
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b0496feb",
+   "metadata": {},
+   "source": [
+    "#### Snakefile Breakdown: \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: for this snakefile breakdown section, I would actually start this when you are first introducing the hello world script so that it flows better.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 17]
```diff
@@ -0,0 +1,711 @@
...
+    "        \"output_1.txt\"\n",
+    "    output: \n",
+    "        \"output_2.txt\" \n",
+    "```\n",
+    "\n",
+    "**Shell Command:** \n",
+    "\n",
+    "The shell keyword is used to specify the shell command that will be executed to produce the output files.\n",
+    "\n",
+    "#### Commandline Command Breakdown: \n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: I would shift this up when you first run hello world with the pcluster plugin.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 18]
```diff
@@ -0,0 +1,711 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f3bc502c",
+   "metadata": {},
+   "source": [
+    "```bash\n",
+    "#Navigate to working directory \n",
+    "mkdir ~/workdir/bioinfo-workflow\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: you can make this into a code cell.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 19]
```diff
@@ -0,0 +1,711 @@
...
+    "\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1b25fd55",
+   "metadata": {},
+   "source": [
+    "### Snakefile Breakdown:\n",
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: I would add this before you make the script do the user can understand what tools you are using and why before they create a script.

File: notebooks/AWS-ParallelCluster.ipynb [hunk 20]
```diff
@@ -0,0 +1,711 @@
...
+    "  * **bwa_index:** Indexes the reference genome file (data/genome.fa) for alignment.\n",
+    "  * **bwa_map:** Maps the sequencing reads (data/samples/{sample}.fastq) to the indexed genome and converts the output to BAM format.\n",
+    "  * **samtools_sort:** Sorts the BAM files generated from the mapping step.\n",
+    "  * **samtools_index:** Indexes the sorted BAM files for faster access.\n",
+    "  * **bcftools_call:** Calls genetic variants from the sorted and indexed BAM files.\n",
+    "  * **plot_quals:** Generates a plot of the quality of the called variants.\n",
+    "  \n",
+    "* The **conda environment** required for each rule is derived from the `conda_env` variable found within the config file\n",
+    "* Each rule uses **shell commands** to perform the required bioinformatics tasks (e.g., bwa index, bwa mem, samtools sort, samtools index, bcftools mpileup, bcftools call).\n",
+    "* Th
```
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: I  dont think there is a need for a breakdown here since you already explained this before. you can just add the details about the --use-conda and -j flag during the 'Execute the Workflow' overview.

File: notebooks/AWS-ParallelCluster.ipynb (hunk 2 above)
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: Also just as a general rule the title should be the font size of 1st heading (h1) so in markdown thats `#` then the main section stated in the checklist (Overview, Learning Objectives, Prerequsities, Pricing, Get Started, Conclusion, Clean up) should be h2 `##` any subsections within them can be h3 (`###`) or h4 (`####').

**Review by zbyosufzai — COMMENTED** [review]
please resolve comments

File: notebooks/AWS-ParallelCluster.ipynb (hunk 20 above)
Comment (zbyosufzai) [inline notebooks/AWS-ParallelCluster.ipynb]: Dont forget to add a Conclusion section before this.

**Review by kyleoconnell-NIH — APPROVED** [review]
